apm-agent-plugins/apm-redis-plugin/apm-lettuce-plugin/src/main/java/co/elastic/apm/agent/redis/lettuce/Lettuce34StopSpanInstrumentation.java

```diff
@@ -0,0 +1,136 @@
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.redis.lettuce;
+
+import co.elastic.apm.agent.bci.VisibleForAdvice;
+import co.elastic.apm.agent.impl.transaction.Span;
+import com.lambdaworks.redis.protocol.RedisCommand;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.NamedElement;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Arrays;
+import java.util.Collection;
+
+import static co.elastic.apm.agent.redis.lettuce.Lettuce34StartSpanInstrumentation.commandToSpan;
+import static net.bytebuddy.matcher.ElementMatchers.declaresMethod;
+import static net.bytebuddy.matcher.ElementMatchers.hasSuperType;
+import static net.bytebuddy.matcher.ElementMatchers.nameContains;
+import static net.bytebuddy.matcher.ElementMatchers.nameStartsWith;
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static net.bytebuddy.matcher.ElementMatchers.takesArguments;
+
+/**
+ * Stops the span created by {@link Lettuce5StartSpanInstrumentation} when one of the following methods are called:
+ * <ul>
+ *     <li>{@link RedisCommand#complete()}</li>
+ *     <li>{@link RedisCommand#completeExceptionally(Throwable)}</li>
+ *     <li>{@link RedisCommand#cancel()}</li>
+ * </ul>
+ */
+public abstract class Lettuce34StopSpanInstrumentation extends Lettuce34Instrumentation {
+
+    @VisibleForAdvice
+    public static final Logger logger = LoggerFactory.getLogger(Lettuce34StopSpanInstrumentation.class);
+
+    @Override
+    public ElementMatcher<? super NamedElement> getTypeMatcherPreFilter() {
+        return nameStartsWith("com.lambdaworks.redis").and(nameContains("Command"));
+    }
+
+    @Override
+    public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+        return hasSuperType(named("com.lambdaworks.redis.protocol.RedisCommand"))
+            // introduced in Lettuce 3.4.0
+            .and(declaresMethod(named("getType")));
+    }
+
+    @Override
+    public Collection<String> getInstrumentationGroupNames() {
+        return Arrays.asList("redis", "lettuce");
+    }
+
+    public static class OnComplete extends Lettuce34StopSpanInstrumentation {
+
+        @Advice.OnMethodEnter(suppress = Throwable.class)
+        private static void beforeComplete(@Advice.This RedisCommand command) {
+            if (!command.isDone() && !command.isCancelled()) {
```

Comment:
Why is this check necessary?

Reply:
A command can be canceled after it's already done. That's also one of the reasons why wrapping and recycling does not work with `RedisCommand`s.

Reply:
But since `commandToSpan` is concurrent, you can only remove once

Reply:
riiight! That should even be safer because it's not a check-than-act.